### ISSUE TITLE
[Bug]: Flush hangs after compact twice

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#12075]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20211118-cdf8e79
- Deployment mode(standalone or cluster): cluster
- SDK version(e.g. pymilvus v2.0.0rc2): 2.0.0rc9.dev11
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others: dataCooed.enableCompaction: true, compacrion.retentionDuration: 20
```

### Current Behavior

1. Deploy cluster milvus 
2. Run test case `test_compact_twice` twice
```
    def test_compact_twice(self):
        """
        target: test compact twice
        method: 1.create with shard_num=1
                2.insert and flush twice (two segments)
                3.compact
                4.insert new data
                5.compact
        expected: Merge into one segment
        """
        # init collection with one shard, insert into two segments
        collection_w = self.collection_insert_two_segments_one_shard(prefix, tmp_nb)

        # first compact two segments
        collection_w.compact()
        collection_w.wait_for_compaction_completed()
        c_plans1 = collection_w.get_compaction_plans()[0]
        target_1 = c_plans1.plans[0].target

        # insert new data
        df = cf.gen_default_dataframe_data(tmp_nb)
        collection_w.insert(df)
        log.debug(collection_w.num_entities)

        # second compact
        collection_w.compact()
        collection_w.wait_for_compaction_completed()
        collection_w.get_compaction_state()
        c_plans2 = collection_w.get_compaction_plans()[0]

        assert target_1 in c_plans2.plans[0].sources
        log.debug(c_plans2.plans[0].target)
``` 
The first run case passed. But the second run flush hangs.

**Client Logs:**
[client.log](https://github.com/milvus-io/milvus/files/7560769/client.log)

**Server Logs:** 
[milvus_logs.tar.gz](https://github.com/milvus-io/milvus/files/7560780/milvus_logs.tar.gz)

### Anything else?

```
spec:
  dependencies:
    etcd:
      inCluster:
        deletionPolicy: Delete
        pvcDeletion: true
    pulsar:
      inCluster:
        deletionPolicy: Delete
        pvcDeletion: true
    storage:
      inCluster:
        deletionPolicy: Delete
        pvcDeletion: true
  components:
    image: harbor.zilliz.cc/milvus/milvus:master-20211118-cdf8e79
    proxy:
      serviceType: LoadBalancer
  config:
    queryNode:
      gracefulTime: 0
    dataCoord:
      enableCompaction: true
      compaction:
        retentionDuration: 20
```

**Comment by binbinlv** [comment]
Similar issue, 

1. Do several times' compact
2. Drop this collection
3. Create new collection, and insert
4. collection.num_entities [Hangs at here]
collection name: test_search_collection_binbin

Logs:
[compact.tar.zip](https://github.com/milvus-io/milvus/files/7560830/compact.tar.zip)

**Comment by ThreadDao** [comment]
Flush hangs when inserted into the 10th segment (auto-merge threshold).

1. create a collection with `shard_num=1`
2. cycle insert and flush 9 times
3. hangs on 10th flush
```
    def test_compact_threshold_auto_merge(self):
        """
        target: test num (segment_size < 1/2Max) reaches auto-merge threshold 10
        method: 1.create with shard_num=1
                2.insert flush 10 times (merge threshold 10)
                3.wait for compaction, load
        expected: Get query segments into to verify segments auto-merged into one
        """
        threshold = 10

        # create collection and insert into threshold segments
        collection_w = self.init_collection_wrap(name=cf.gen_unique_str(prefix), shards_num=1)
        for i in range(threshold):
            df = cf.gen_default_dataframe_data(1)
            collection_w.insert(df)
            assert collection_w.num_entities == i+1

        # todo compaction cost
        collection_w.get_compaction_plans()
        collection_w.load()
        segments_info = self.utility_wrap.get_query_segment_info(collection_w.name)[0]

        # verify segments reaches threshold, auto-merge ten segments into one
        assert len(segments_info) == 1
```
**Note:** If you don't reproduce it the first time, then you run the case again

**Comment by XuanYang-cn** [comment]
The flush hangs because:

As the 10th segment `SaveBinlogPath`, datacoord will trigger auto compaction immediately. And the compaction was completed before the Next GetPersistentSegmentInfo req from SDK.

Current flush procedure in SDK:

```python
segIDs = flush()  # At this time, the 10th segment is returned

## At this time, the compacted segment is returned, thus stucked here
wait until all segID in segIDs shows  in  `GetPersistentSegmentInfo`
```

**Comment by XuanYang-cn** [comment]
Candidate options: update 
```
message PersistentSegmentInfo {
  int64 segmentID = 1;
  int64 collectionID = 2;
  int64 partitionID = 3;
  int64 num_rows = 4;
  common.SegmentState state = 5;
  bool createdByCompaction = 6;  // new
  repeated int64 compactedFrom = 7; // new
}
```

in SDK,  take care compacted segments in GetPersistentSegmentInfo of `_wait_for_flushed`

**Comment by XuanYang-cn** [comment]
> Candidate options: update
> 
> ```
> message PersistentSegmentInfo {
>   int64 segmentID = 1;
>   int64 collectionID = 2;
>   int64 partitionID = 3;
>   int64 num_rows = 4;
>   common.SegmentState state = 5;
>   bool createdByCompaction = 6;  // new
>   repeated int64 compactedFrom = 7; // new
> }
> ```
> 
> in SDK, take care compacted segments in GetPersistentSegmentInfo of `_wait_for_flushed`

This solution cannot handle a multi-compacted segment, which generates from a compacted segment.

**Comment by xiaofan-luan** [comment]
please remember to notify the SDK maintainer

**Comment by sunby** [comment]
@yhmo @XuanYang-cn I will add a new interface `GetFlushState` which receives a slice of segment id and return true only if all segments are flushed. This interface is used to check flush progress after calling `Flush`

**Comment by sunby** [comment]
@ThreadDao Fixed, please check it.
/unassign @sunby 
/assign @ThreadDao

**Comment by ThreadDao** [comment]
Fixed `master-20211124-7b870e3`
